### PR TITLE
[DiffusionGemma] Cast the decoder padding mask to bool

### PR DESCRIPTION
**Pull request by kashif**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47295)

`create_diffusion_decoder_attention_mask` calls the low-level mask interface directly. That one keeps the dtype of the 2D mask it is given, unlike the high-level mask creation functions which cast it. So a `1`/`0` integer padding mask, like the one a tokenizer returns, ends up propagating its dtype and we get an int64 mask back instead of a boolean one, which then blows up in sdpa.

The existing mask tests all pass a `torch.bool` mask, so this never showed up. Added a test with an integer mask, it fails on main and passes here.

Found while training DiffusionGemma from TRL, which builds its `decoder_attention_mask` with `torch.long`.

### CODE REVIEW

File: src/transformers/models/diffusion_gemma/modeling_diffusion_gemma.py [hunk 1]
```diff
@@ -1392,6 +1392,10 @@ def create_diffusion_decoder_attention_mask(
         ):
             return decoder_attention_mask
 
+        # Contrarily to the high-level mask creation functions, the mask interface used below does not cast the 2D
+        # mask, and an integer one would propagate its dtype to the final mask instead of yielding a boolean mask
+        decoder_attention_mask = decoder_attention_mask.bool()
```
Comment (zucchini-nlp) [inline src/transformers/models/diffusion_gemma/modeling_diffusion_gemma.py]: could you move the casting under a cond, if the mask is 2D?

Copied from masking utils:
```python
# Move the mask to correct device, and potentially switch dtype for efficiency
    if attention_mask is not None and attention_mask.ndim == 2:
        attention_mask = attention_mask.to(device=inputs_embeds.device, dtype=torch.bool)
```

**Review by zucchini-nlp — APPROVED** [review]
Nice catch!
